### PR TITLE
caddytls: add TLS 1.3 support

### PR DESCRIPTION
**Pull request by crvv**
### 1. What does this change do, exactly?
add TLS 1.3

### 2. Please link to the relevant issues.
#2080 

### 3. Which documentation changes (if any) need to be made because of this PR?
https://caddyserver.com/docs/tls
1. "tls1.3" is a new suppoted protocol if caddy is built by Go 1.12 or newer.
2. TLS 1.3 cipher suites are not configurable.

### 4. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I am willing to help maintain this change if there are issues with it later

If caddy should be able to be built by Go 1.11 after the releasing of 1.12, I think this PR can be accepted now.
If not, the new file with build tag "go1.12" is useless and the change can be applied to config.go directly when Go 1.12 is released.

### CODE REVIEW

**Comment by mholt** [comment]
Nice, thanks for jumping on this. (Realizing that, of course, this is a temporary solution, as once Go 1.12 is released, we can just update it in-place.)

While we're at it, can we remove the CBC ciphers from the defaultCiphers lists? I think we should take this opportunity to modernize Caddy's TLS config to be up to snuff with the latest: TLS 1.2 and 1.3.

**Comment by crvv** [comment]
I tested this on SSL Labs. The result is https://www.ssllabs.com/ssltest/analyze.html?d=tw.crvv.me&hideResults=on

With Let's Encrypt RSA certificates, the failed clients are `IE 11 on Windows 7/8` and `Safari 6/7/8`

Is that OK?

**Comment by abiosoft** [comment]
I do not think this should be an issue.

> I tested this on SSL Labs. The result is https://www.ssllabs.com/ssltest/analyze.html?d=tw.crvv.me&hideResults=on
>
> With Let's Encrypt RSA certificates, the failed clients are IE 11 on Windows 7/8 and Safari 6/7/8
>
> Is that OK?

**Comment by elcore** [comment]
> While we're at it, can we remove the CBC ciphers from the defaultCiphers lists? I think we should take this opportunity to modernize Caddy's TLS config to be up to snuff with the latest: TLS 1.2 and 1.3.

We definitely should add TLS 1.3 cipher suites 😄

**Comment by crvv** [comment]
@elcore 
TLS 1.3 ciphers are not configurable via `Config.CipherSuites`.
Please see golang/go#29349 for details.

**Comment by elcore** [comment]
> @elcore
TLS 1.3 ciphers are not configurable via Config.CipherSuites.
Please see golang/go#29349 for details.

Oh... Thank you!

**Comment by brink668** [comment]
Is there an ETA on this?

**Comment by mholt** [comment]
To use Caddy with TLS 1.3 today, go ahead and build from @crvv's branch -- and when Go 1.12 comes out next month, this PR should be updated to make TLS 1.3 the default (without the build constraint which it currently has), and then we'll merge in the PR.

**Comment by mholt** [comment]
@crvv Will we need to set an env variable to enable TLS 1.3 by default? https://github.com/golang/go/issues/30055

**Comment by crvv** [comment]
We need to set that.
But let's wait for some responses to my comments in that issue.

**Comment by mholt** [comment]
@crvv Great! Looks like a doc change will be merged, and I'm good with it if we include that Setenv in this PR to enable it by default.

**Comment by Henrocker** [comment]
Made some tests today, compiling caddy with go1.12rc1 and [this](https://hnrk.io/md/caddy.patch.html) patch. 
Result: TLS 1.3 did not work anymore.

Switched then back to go1.12beta2 and applied the same patch and boom, TLS 1.3 works again... Wonder where that comes from.

**Comment by mholt** [comment]
@Henrocker I think it's because rc1 now requires an environment variable to be set. See the issue I linked to in a comment above. Would love it if you could help figure out if that's what is needed!

**Comment by Henrocker** [comment]
Thanks for pointing me out. Gonna do some further tests tomorrow and will report back if the envvar fixes this. Thanks

**Comment by Henrocker** [comment]
Applied [patch](https://hnrk.io/md/caddy.patch) and Caddy with go1.12rc1 Build command:
```
$ GODEBUG=tls13=1 go run build.go
```

Command to run Caddy:
```
$ GODEBUG=tls13=1 ./caddy
```

TLS 1.3 works! :-)

**Comment by mholt** [comment]
@Henrocker Any chance we can set that env var in code instead of the user having to do it?

**Comment by Henrocker** [comment]
Maybe something like https://github.com/golang/go/issues/30055#issuecomment-463146031 could be done, but since my understanding in golang is limited, I'm afraid I can only do some tests, if there's something new to try out.

File: caddytls/setup.go [hunk 1]
```diff
@@ -34,6 +34,9 @@ import (
 )
 
 func init() {
+	// opt-in TLS 1.3 for Go1.12
+	os.Setenv("GODEBUG", os.Getenv("GODEBUG")+",tls13=1")
```
Comment (crvv) [inline caddytls/setup.go]: @Henrocker  @mholt 
I have added this env in code.

Comment (whalehub) [reply caddytls/setup.go]: @mholt I can confirm that compiling Caddy with go1.12rc1 and @crvv's change successfully enables TLS 1.3 support in Caddy without the need to manually set an env variable in the OS.

Comment (mholt) [reply caddytls/setup.go]: Excellent!

Can we put a TODO in the comment so we remember to revisit this later (i.e. after Go 1.13)?

Comment (crvv) [reply caddytls/setup.go]: Yes, I have added the TODO.

**Comment by mholt** [comment]
Go 1.12 is released; just waiting for CI systems to pick up the new version.

**Comment by mholt** [comment]
AppVeyor is refusing to pick up the merge that upgrades to Go 1.12, but the tests on Travis are passing, so I'm merging this sucker.

Thanks everyone!! How exciting.

**Comment by kleveralt** [comment]
tls1.0 and tls1.2 is no longer available?

I use this config but only tls1.2 and tls1.3 are supported, since this commit...
protocols tls1.0 tls1.3

**Comment by whitestrake** [comment]
The documentation does state that "Supported protocols and default protocol versions may be changed at any time", but if that's the case, it should be updated at the next opportunity to reflect the current available protocols.

**Comment by mholt** [comment]
TLS 1.0 and 1.1 are supported but not enabled by default, you have to enable those manually, but I highly, highly, strongly, super duper discourage that.

**Comment by kleveralt** [comment]
I understand, but ssllabs.com replys only tls1.2 and 1.3 are supported with this cofig,
    tls {
        protocols tls1.0 tls1.3
    }

image

**Comment by mholt** [comment]
Oh, I see what you mean. Hmm. Can you open a new issue please?

**Comment by crvv** [comment]
With this PR, there are only AEAD ciphers in the default cipher suites.
AEAD is new in TLS 1.2
So TLS 1.0 and 1.1 won't work with default cipher suites.

If you want TLS 1.0 and 1.1, you must specify cipher suites you want in the Caddyfile.

**Comment by mholt** [comment]
Annnd we have a winner, totally forgot about that. 😅  That would probably be the reason! Thanks @crvv.

**Pull request merged**